Manage AWS Nat Gateways

linuxdynasty opened the pull request:
##### Issue Type:
 
- New Module Pull Request

##### Plugin Name:

ec2_vpc_nat_gateway

##### Ansible Version:

```
ansible 2.1.0
```

##### Summary:
I based this module off of @Etherdaemon original pull request.  I closed my original PR due to me including to many commits in my last PR. #1797
    Manage AWS Nat Gateways.
    * Create an AWS Nat Gateway.
    * Delete an AWS Nat Gateway.
    * If Nat Gateway exist in subnet and the option is passed to not create one, it will then return the Nat Gateway object.

```
- name: Launching NAT Gateway and allocate a new eip.
  ec2_vpc_nat_gateway:
    region: us-west-2
    state: present
    subnet_id: "{{ test_subnet_id }}"
    wait: yes
    wait_timeout: 600
  register: nat

- debug:
    var: nat
- fail:
    msg: "Failed to create"
  when: '"{{ nat["changed"] }}" != "True"'

- name: Launch a new gateway only if one does not exist already in this subnet.
  ec2_vpc_nat_gateway:
    if_exist_do_not_create: yes
    region: us-west-2
    state: present
    subnet_id: "{{ test_subnet_id }}"
    wait: yes
    wait_timeout: 600
  register: nat_idempotent

- debug:
    var: nat_idempotent
- fail:
    msg: "Failed to be idempotent"
  when: '"{{ nat_idempotent["changed"] }}" == "True"'

- name: Launching NAT Gateway and allocate a new eip even if one already exists in the subnet.
  ec2_vpc_nat_gateway:
    region: us-west-2
    state: present
    subnet_id: "{{ test_subnet_id }}"
    wait: yes
    wait_timeout: 600
  register: new_nat

- debug:
    var: new_nat
- fail:
    msg: "Failed to create"
  when: '"{{ new_nat["changed"] }}" != "True"'

- name: Launching NAT Gateway with allocation id, this call is idempotent and will not create anything.
  ec2_vpc_nat_gateway:
    allocation_id: eipalloc-1234567
    region: us-west-2
    state: present
    subnet_id: "{{ test_subnet_id }}"
    wait: yes
    wait_timeout: 600
  register: nat_with_eipalloc

- debug:
    var: nat_with_eipalloc
- fail:
    msg: 'Failed to be idempotent.'
  when: '"{{ nat_with_eipalloc["changed"]  }}" == "True"'

- name: Delete the 1st nat gateway and do not wait for it to finish
  ec2_vpc_nat_gateway:
    region: us-west-2
    nat_gateway_id: "{{ nat.nat_gateway_id }}"
    state: absent

- name: Delete the nat_with_eipalloc and release the eip
  ec2_vpc_nat_gateway:
    region: us-west-2
    nat_gateway_id: "{{ new_nat.nat_gateway_id }}"
    release_eip: yes
    state: absent
    wait: yes
    wait_timeout: 600
```

##### Example output:
This is deleting a nat gateway.
```
***********************************
PARSED OUTPUT
{
    "changed": true,
    "invocation": {
        "module_args": {
            "allocation_id": null,
            "aws_access_key": null,
            "aws_secret_key": null,
            "client_token": null,
            "ec2_url": null,
            "eip_address": null,
            "if_exist_do_not_create": false,
            "nat_gateway_id": "nat-0b2f9f2ac3f51a653",
            "profile": null,
            "region": null,
            "release_eip": true,
            "security_token": null,
            "state": "absent",
            "subnet_id": null,
            "validate_certs": true,
            "wait": true,
            "wait_timeout": 320
        }
    },
    "result": {
        "create_time": "2016-03-05T05:19:20.282000+00:00",
        "delete_time": "2016-03-05T22:07:29.120000+00:00",
        "nat_gateway_addresses": [
            {
                "allocation_id": "eipalloc-12345",
                "network_interface_id": "eni-12345",
                "private_ip": "10.0.0.100",
                "public_ip": "52.52.52.52"
            }
        ],
        "nat_gateway_id": "nat-0b2f9f2ac3f51a653",
        "state": "deleted",
        "subnet_id": "subnet-12345",
        "vpc_id": "vpc-12345"
    }
}
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ansible/ansible-modules-extras/1882)
<!-- Reviewable:end -->

gregdek commented:
Thanks @linuxdynasty for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

linuxdynasty commented:
@mingchuno Also please make sure to pip install boto as well. If that works, that means the ec2.py module is not adhering to the boto3=True.

mingchuno commented:
@linuxdynasty I have pip installed both boto and boto3, I am using OSX. But anyway, let me check again tmr. Maybe it is my bad.

mingchuno commented:
@linuxdynasty yup, tried but still fail

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/mcor/.ansible/tmp/ansible-tmp-1458619207.79-93872800676393/ec2_vpc_nat_gateway", line 3113, in <module>
    main()
  File "/Users/mcor/.ansible/tmp/ansible-tmp-1458619207.79-93872800676393/ec2_vpc_nat_gateway", line 878, in main
    get_aws_connection_info(module, boto3=True)
  File "/Users/mcor/.ansible/tmp/ansible-tmp-1458619207.79-93872800676393/ec2_vpc_nat_gateway", line 3022, in get_aws_connection_info
    region = boto.config.get('Boto', 'aws_region')
NameError: global name 'boto' is not defined
```

I have pip installed both `boto` and `boto3`. I am in OSX with ansible 2.0.1.0

Etherdaemon commented:
@mingchuno @linuxdynasty  I think it's because you dont have a region set in your play. It looks like a shortcoming of the ec2.py but either way you should actually be setting your region in either your play or environment variables. It was updated/fixed in devel so later version will not have this issue.

linuxdynasty commented:
@Etherdaemon Thanks for the suggestion and I think you maybe right. @mingchuno please make sure you have your AWS Environment settings correct... Example..

```
AWS_ACCESS_KEY_ID=super secret access key
AWS_DEFAULT_REGION=us-west-2
AWS_SECRET_ACCESS_KEY=super secret key goes here
```

mingchuno commented:
```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/mcor/.ansible/tmp/ansible-tmp-1458710531.43-101246038100876/ec2_vpc_nat_gateway", line 3247, in <module>
    main()
  File "/Users/mcor/.ansible/tmp/ansible-tmp-1458710531.43-101246038100876/ec2_vpc_nat_gateway", line 905, in main
    client_token
  File "/Users/mcor/.ansible/tmp/ansible-tmp-1458710531.43-101246038100876/ec2_vpc_nat_gateway", line 751, in pre_create
    allocation_id = get_eip_allocation_id_by_address(client)
TypeError: get_eip_allocation_id_by_address() takes at least 2 arguments (1 given)
```

the call is around line 751

```
    elif eip_address or allocation_id:
        if eip_address and not allocation_id:
            allocation_id = get_eip_allocation_id_by_address(client)
```

my playbook is sth like this:

```
- name: create NAT gateway
  ec2_vpc_nat_gateway:
    state: present
    subnet_id: "{{ item[1].id }}"
    eip_address: "{{ item[0].public_ip }}"
    region: "{{  region  }}"
    wait: yes
  with_together:
    - "{{ nat_eip.results }}"
    - "{{ public_subnet_fact.subnets }}"
```

linuxdynasty commented:
@mingchuno thank you for finding that bug, it was missing a parameter. I pushed a fix for that.

JustinHead commented:
I'm getting this traceback if no nat gateway yet exists and setting if_exist_do_not_create: true. Need to set a default for existing_gateways in get_nat_gateways().

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/home/jhead/.ansible/tmp/ansible-tmp-1458927714.75-263444293548113/ec2_vpc_nat_gateway", line 3117, in <module>
    main()
  File "/home/jhead/.ansible/tmp/ansible-tmp-1458927714.75-263444293548113/ec2_vpc_nat_gateway", line 909, in main
    client_token
  File "/home/jhead/.ansible/tmp/ansible-tmp-1458927714.75-263444293548113/ec2_vpc_nat_gateway", line 741, in pre_create
    gateway_in_subnet_exists(client, subnet_id)
  File "/home/jhead/.ansible/tmp/ansible-tmp-1458927714.75-263444293548113/ec2_vpc_nat_gateway", line 480, in gateway_in_subnet_exists
    gws_retrieved, _, gws = get_nat_gateways(client, subnet_id)
  File "/home/jhead/.ansible/tmp/ansible-tmp-1458927714.75-263444293548113/ec2_vpc_nat_gateway", line 369, in get_nat_gateways
    return gateways_retrieved, err_msg, existing_gateways
UnboundLocalError: local variable 'existing_gateways' referenced before assignment
```

linuxdynasty commented:
@JustinHead thank you for the catch. I am going to write tests for this module today and let you know once all is resolved.

JustinHead commented:
Actually, it would appear my issues are due to a boto3 bug with assumed roles in ansible.

https://github.com/ansible/ansible/issues/13727

linuxdynasty commented:
@justingood  and @mingchuno I added unit and integration test to this PR and you can see the results of the test on my repo https://github.com/linuxdynasty/ld-ansible-modules

sgzijl commented:
FYI...

I get a stack trace when exceeding the limit of 5 NAT gateways per zone:

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/foo/.ansible/tmp/ansible-tmp-1460148497.43-137096035157791/ec2_vpc_nat_gateway", line 3252, in <module>
    main()
  File "/Users/foo/.ansible/tmp/ansible-tmp-1460148497.43-137096035157791/ec2_vpc_nat_gateway", line 1040, in main
    client_token, check_mode=check_mode
  File "/Users/foo/.ansible/tmp/ansible-tmp-1460148497.43-137096035157791/ec2_vpc_nat_gateway", line 876, in pre_create
    wait, wait_timeout, if_exist_do_not_create, check_mode=check_mode
  File "/Users/foo/.ansible/tmp/ansible-tmp-1460148497.43-137096035157791/ec2_vpc_nat_gateway", line 763, in create
    return success, changed, err_msg, result
UnboundLocalError: local variable 'result' referenced before assignment

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "ec2_vpc_nat_gateway"}, "parsed": false}
```

When I set result to an empty string on line 725, it returns a valid error message:

```
msg: An error occurred (NatGatewayLimitExceeded) when calling the CreateNatGateway operation: Performing this operation would exceed the limit of 5 NAT gateways
```

linuxdynasty commented:
@sgzijl, thank you for catching that. I will update shortly as well as update my tests.

linuxdynasty commented:
@sgzijl, I updated the module, by setting the default result to None when it catches the exception.

linuxdynasty commented:
@mingchuno , @Etherdaemon , @JustinHead, @aioue and @sgzijl, if you guys are using or have tested this module, can you please give updated feedback and if this can include a shipit?
@khogeland this module does take the eip as an option.

gregdek commented:
Thanks @linuxdynasty for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

wyaeld commented:
is `if_exist_do_not_create` a pre-existing name from somewhere, because its a very awkward construction, particulary when `if_exist_do_not_create: false`, now you have a double-negative.

If well written ansible modules are supposed to be idempotent, why is this needed?

aioue commented:
@linuxdynasty I have tested a possible duplicate module effort https://github.com/ansible/ansible-modules-extras/pull/1687.

linuxdynasty commented:
@wyaeld the if_exist_do_not_create key is due to the fact the AWS Nat Gateways does not support Tags. So the only way to validate that a Gateway already exist, is by using the ARN of the gateway or the EIP that is attached to a gateway. This option allows the user  to not have to hard code the ARN in the module, especially if they are using this module to create the gateway. Also this option is very helpful, if you do not want to create more than one Nat Gateway in a subnet. I am more than happy to rename this option.

gregdek commented:
@linuxdynasty A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@linuxdynasty Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@linuxdynasty A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

linuxdynasty commented:
@gregdek, this has not been flagged or commented by anyone except for the buildbot. All checks are passing.

gregdek commented:
Kicking. Still waiting for shipits.

gregdek commented:
Thanks @linuxdynasty for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

sgzijl commented:
Module tested in different environments and it works great.

@linuxdynast thanks a million - shipit

wimnat commented:
@Etherdaemon can you close this PR in favor of https://github.com/ansible/ansible-modules-extras/pull/1687 as I believe you both worked on that one any way?

linuxdynasty commented:
@wimnat I am the who worked on this PR on top of @Etherdaemon work. This module contains an added feature that #1687 does not, which is to search by EIP as well as this PR contains Tests with it. Either way, I am fine on how the Ansible team and community would like to proceed.

gregdek commented:
After discussion, it's time to make a decision, and the decision I'm making is that we're going with this PR. @jonhadfield and @Etherdaemon, thanks to both of you for your help and your patience throughout this process.

gundalow commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,1070 @@
+  allocation_id:
+    description:
+      - The id of the elastic IP allocation. If this is not passed and the
+        eip_address is not passed. An EIP is generated for this Nat Gateway
```
Please add a full stop at the end of this sentence. 

Also `NAT` rather than `Nat`?

gundalow commented:
s/Nat/NAT/g in the documentation please. Dito for IP.

gundalow commented:
We are happy for the docs updates to be done after merge

gregdek commented:
Marking "shipit" for final review. Thanks @linuxdynasty and all who have pursued this one for so long.

alikins commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,1070 @@
+    results = dict()
+    if isinstance(data, dict):
+        for key, val in data.items():
+            key = re.sub(r'(([A-Z]{1,3}){1})', r'_\1', key).lower()
```
This works as 're' is import from basic, but it's preferred to add a 'import re' for clarity

mattclay commented:
@linuxdynasty Once this is merged, please submit a PR to add the integration tests to the ansible repo.

gundalow commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,1070 @@
+options:
+  state:
+    description:
+      - Ensure NAT Gateway is present or absent
```
Please add a full stop at the end of this sentence.

alikins commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,1070 @@
+                )
+            )
+            if gws_retrieved and nat_gateway:
+                nat_gateway = nat_gateway[0]
```
This is a little odd to me. Esp since the first nat_gateway is a list and the second is a dict.
Maybe 'nat_gateways' and 'nat_gateway' ?

alikins commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,1070 @@
+                    break
+
+                elif nat_gateway.get('state') == 'pending':
+                    if nat_gateway.has_key('failure_message'):
```
cloud/amazon/ec2_vpc_nat_gateway.py:452:34: W601 .has_key() is deprecated, use 'in'

Either works, but...

alikins commented:
LGTM. I noted a few non-critical things, but nothing that would block a merge.

alikins commented:
And +1000 for tests.

linuxdynasty commented:
@gundalow and @alikins, I have cleaned up the documentation and removed the has_key in favor for in.

ryansb commented:
I'm getting some odd behavior when there aren't any allocations - the module succeeds, but gives an error. 

```
ok: [localhost] => {
    "nat": {
        "changed": "False", 
        "msg": "An error occurred (AddressLimitExceeded) when calling the AllocateAddress operation: The maximum number of addresses has been reached.", 
        "success": false
    }
}
```

I'll look for an exact cause and see if I can fix it on my end.

ryansb commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,1078 @@
+            )
+
+    if not success:
+        module.exit_json(
```
This needs to be `fail_json` instead

linuxdynasty replied:
Thank you for catching this. I just pushed the change.

sgzijl commented:
@ryansb - the maximum number of NAT gateways in a VPC is 5 per AZ. Also; each AWS account has a limit of 5 EIP's (Elastic IP addresses) by default. I think your AWS account is out of EIP's...

ryansb commented:
@sgzijl I know why it failed, the issue is that the task returned success, so when I looked at the playbook summary it showed up as `changed` rather than `failed`. 

```
PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0 
```

ryansb commented:
Other than that, works great once I give it an available allocation. One minor thing is that it doesn't deallocate its EIP, it only dissociates it. I'm ok with that behavior, but would like to get other folks' opinions on whether that's the behavior they want.

sgzijl commented:
@ryansb Ah ok, good catch.

My opinion about disassociation - I'm ok with that behavior also. Once given away an EIP, you will never get it back (may cause problems in many environments). Plus, behavior same as `ec2_eip`it's default (but without the `release_on_disassociation` option)

linuxdynasty commented:
@ryansb and @sgzijl , do you guys think I should add the release_on_disassociation option?

linuxdynasty commented:
I just fixed the error message when releasing an eip from a NAT gateway that is in a transition state. I was ignoring the error from release_eip. In order for an EIP to get released from a NAT gateway, one of the following has to occur.. (I updated the documentation string in the module)
- Nat gateway has to be in a deleted state (use wait=yes and wait_timeout=300 while passing the absent state) OR
- Nat gateway has to be in a deleted state already.

Other wise if you delete and do not wait. The module will now fail because it can not release the EIP while the gateway is in transition.

``` yaml
- name: Delete nat gateway and release EIP.
    ec2_vpc_nat_gateway:
      state: absent
      nat_gateway_id: nat-12345678
      release_eip: yes
      wait: yes
      wait_timeout: 300
      region: ap-southeast-2
```

ryansb commented:
Oh, I missed the `release_eip` option. Thanks for adding that to the docs, this looks great!

aioue commented:
what version is this targetted for official release in?

linuxdynasty commented:
@aioue The 2.2 release if I am not mistaken.

ryansb commented:
Yup, that's right - this'll be part of the 2.2 release of Ansible, targeted for this September.

RichardN commented:
I notice that when a nat gateway is being created it seems to return the gateway details as CamelCase, but when one is already in the environment (with "if_exist_do_not_create: yes"), it returns the data after passing through convert_to_lower.   This means working with the module gets very messy...

wimnat commented:
@RichardN - create a new issue for this and ill submit a PR

linuxdynasty commented:
@RichardN, thank you for that catch. @wimnat if you got this, great. Could you also update the tests to validate that all results are returned in lower_case. If not I can do it.

linuxdynasty commented:
Matter of fact I already have a PR open for adding support for the retry decorator. I will go ahead and add the fix there @wimnat and @RichardN #2721

linuxdynasty commented:
Fixed in #2721
